Library/Homebrew/utils/popen.rb

```diff
@@ -1,46 +1,70 @@
-    output = popen_read(*args, **options, &block)
-    return output if $CHILD_STATUS.success?
+  module Popen
+    DEFAULT_BUFFER_SIZE = 4096
+    private_constant :DEFAULT_BUFFER_SIZE
 
-    raise ErrorDuringExecution.new(args, status: $CHILD_STATUS, output: [[:stdout, output]])
-  end
+    def popen_read(*args, **options, &block)
+      popen(args, "rb", options, &block)
+    end
 
-  def self.popen_write(*args, **options, &block)
-    popen(args, "wb", options, &block)
-  end
+    def safe_popen_read(*args, **options, &block)
+      output = popen_read(*args, **options, &block)
+      return output if $CHILD_STATUS.success?
 
-  def self.safe_popen_write(*args, **options, &block)
-    output = popen_write(*args, **options, &block)
-    return output if $CHILD_STATUS.success?
+      raise ErrorDuringExecution.new(args, status: $CHILD_STATUS, output: [[:stdout, output]])
+    end
 
-    raise ErrorDuringExecution.new(args, status: $CHILD_STATUS, output: [[:stdout, output]])
-  end
+    def popen_write(*args, **options)
+      popen(args, "w+b", options) do |pipe|
+        output = ""
 
-  def self.popen(args, mode, options = {})
-    IO.popen("-", mode) do |pipe|
-      if pipe
-        return pipe.read unless block_given?
+        # Before we yield to the block, capture as much output as we can
+        loop do
+          output += pipe.read_nonblock(DEFAULT_BUFFER_SIZE)
+        rescue IO::WaitReadable, EOFError
+          break
+        end
 
         yield pipe
-      else
-        options[:err] ||= :close unless ENV["HOMEBREW_STDERR"]
-        begin
-          exec(*args, options)
-        rescue Errno::ENOENT
-          $stderr.puts "brew: command not found: #{args[0]}" unless options[:err] == :close
-          exit! 127
-        rescue SystemCallError
-          $stderr.puts "brew: exec failed: #{args[0]}" unless options[:err] == :close
-          exit! 1
+        pipe.close_write
+        IO.select([pipe])
+
+        # Capture the rest of the output
+        output += pipe.read
+        output.freeze
+      end
+    end
+
+    def safe_popen_write(*args, **options, &block)
+      output = popen_write(*args, **options, &block)
+      return output if $CHILD_STATUS.success?
+
+      raise ErrorDuringExecution.new(args, status: $CHILD_STATUS, output: [[:stdout, output]])
+    end
+
+    def popen(args, mode, options = {})
+      IO.popen("-", mode) do |pipe|
+        if pipe
+          return pipe.read unless block_given?
+
+          yield pipe
+        else
+          options[:err] ||= :close unless ENV["HOMEBREW_STDERR"]
+          begin
+            exec(*args, options)
+          rescue Errno::ENOENT
+            $stderr.puts "brew: command not found: #{args[0]}" unless options[:err] == :close
+            exit! 127
+          rescue SystemCallError
+            $stderr.puts "brew: exec failed: #{args[0]}" unless options[:err] == :close
+            exit! 1
+          end
         end
       end
     end
   end
+
+  extend Popen
```

Comment:
Feels a bit weird to define the module and then immediately add it to a wider namespace?

Reply:
That was to preserve the `Utils.popen_foo` syntax (rather than the somewhat awkward `Utils::Popen.popen_foo`).

Maybe we can extract the module outside `Utils`, for example to `Homebrew::Open`.
Then we can call the methods like so:

```
Open.popen_foo(…)
```

What do you think?

Reply:
I'd be tempted to just leave modules as-is from before?

Reply:
Not sure where to put the constant then without bleeding into namespaces.

Reply:
Doesn't matter if the constant bleeds in. If you really are concerned: can make it a named variable instead.

Reply:
Fair enough 👍